manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java

```diff
@@ -0,0 +1,300 @@
+         return;
+      }
+      var cache = getOrCreateMap(header.getCacheName());
+
+      try {
+         cache.put(new String(key), new WrappedByteArray(value));
+         writeSuccess(header);
+      } catch (Throwable t) {
+         writeException(header, t);
+      }
+   }
+
+   public void get(GingersnapHeader header, byte[] key) {
+      var cache = getOrCreateMap(header.getCacheName());
+
+      try {
+         WrappedByteArray wba = cache.getIfPresent(new String(key));
+         if (wba == null) {
+            writeNotExist(header);
+         } else {
+            writeResponse(header.encoder().valueResponse(header, null, channel, OperationStatus.Success, wba.getBytes()));
+         }
+      } catch (Throwable t) {
+         writeException(header, t);
+      }
+   }
+
+   public void putIfAbsent(GingersnapHeader header, byte[] key, byte[] value) {
+      if (header.hasFlag(ProtocolFlag.ForceReturnPreviousValue)) {
+         writeException(header, new UnsupportedOperationException("previous value is not supported!"));
+         return;
+      }
+      var cache = getOrCreateMap(header.getCacheName());
+
+      try {
+         WrappedByteArray wba = cache.asMap().putIfAbsent(new String(key), new WrappedByteArray(value));
+         if (wba == null) {
+            this.writeSuccess(header);
+         } else {
+            writeResponse(header.encoder().emptyResponse(header, null, channel, OperationStatus.OperationNotExecuted));
+         }
+      } catch (Throwable t) {
+         writeException(header, t);
+      }
+   }
+
+   public void remove(GingersnapHeader header, byte[] key) {
+      if (header.hasFlag(ProtocolFlag.ForceReturnPreviousValue)) {
+         writeException(header, new UnsupportedOperationException("previous value is not supported!"));
+         return;
+      }
+      var cache = getOrCreateMap(header.getCacheName());
+
+      try {
+         var prev = cache.asMap().remove(new WrappedByteArray(key));
+         if (prev == null) {
+            this.writeNotExist(header);
+         } else {
+            this.writeSuccess(header);
+         }
+      } catch (Throwable t) {
+         writeException(header, t);
+      }
+   }
+
+   public void ping(GingersnapHeader header) {
+      this.writeResponse(header.encoder().pingResponse(header, null, this.channel, OperationStatus.Success));
+   }
+
+   public void size(GingersnapHeader header) {
+      var cache = maps.getMaps().get(header.getCacheName());
+      var size = cache == null ? -1 : cache.estimatedSize();
+      this.writeResponse(header.encoder().unsignedLongResponse(header, null, this.channel, size));
+   }
+
+   public void putAll(GingersnapHeader header, Map<byte[],byte[]> entryMap) {
+      var cache = getOrCreateMap(header.getCacheName());
+
+      try {
+         for (Map.Entry<byte[], byte[]> entry : entryMap.entrySet()) {
+            cache.put(new String(entry.getKey()), new WrappedByteArray(entry.getValue()));
+         }
```

Comment:
I had debated about using putAll, but even that is not an atomic operation. The main concern is the putAll would require creating yet another map and the only benefit it would give is the presize operation if the map is very large, which we will probably never hit.